pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -1127,14 +1126,15 @@ func newPodNominator(podLister listersv1.PodLister) *nominator {
 func MakeNextPodFunc(queue SchedulingQueue) func() *framework.QueuedPodInfo {
 	return func() *framework.QueuedPodInfo {
 		podInfo, err := queue.Pop()
-		if err == nil {
+		if err == nil && podInfo != nil {
 			klog.V(4).InfoS("About to try and schedule pod", "pod", klog.KObj(podInfo.Pod))
 			for plugin := range podInfo.UnschedulablePlugins {
 				metrics.UnschedulableReason(plugin, podInfo.Pod.Spec.SchedulerName).Dec()
 			}
 			return podInfo
+		} else if err != nil {
```

Comment:
> but better not to bury the reason

I agree. but even use the above logic you write, the `errQueueClosed ` may not be logged because the `queue.Pop()` not always return err when scheduler exit. it depends on the execute order of the code below.

```
// Run begins watching and scheduling. It starts scheduling and blocked until the context is done.
func (sched *Scheduler) Run(ctx context.Context) {
	sched.SchedulingQueue.Run()

	// We need to start scheduleOne loop in a dedicated goroutine,
	// because scheduleOne function hangs on getting the next item
	// from the SchedulingQueue.
	// If there are no new pods to schedule, it will be hanging there
	// and if done in this goroutine it will be blocking closing
	// SchedulingQueue, in effect causing a deadlock on shutdown.
	go wait.UntilWithContext(ctx, sched.scheduleOne, 0)

	<-ctx.Done()
	sched.SchedulingQueue.Close()
}
```

If `SchedulingQueue.Close()` execute first and then scheduleOne executed then there will be a queue closed error. but if the execute order is reversed, there will be no error.

Reply:
I don't think so, scheduleOne will block until we close the scheduleQueue and wake the condition. Anyway, what we want here is not emitting errors causing confusion, if we aim to log the closing queue message, we should log at where we call the `Close` function instead.

Reply:
you are right. what we met is that the process may exited before `scheduleOne` finally execute and then the closing queue message will not be logged.

Reply:
In general, stick to the suggestion here https://github.com/kubernetes/kubernetes/pull/116583/files#r1141797007.